File: extract_data.js
const fs = require('fs');

const html = fs.readFileSync('dormant_btc.html', 'utf8');

// BitInfoCharts dormant addresses page has tables with id="tblOne" and "tblOne2"
const tables = [];
const tableIds = ['tblOne', 'tblOne2'];

tableIds.forEach(id => {
  const tableMatch = html.match(new RegExp(`<table[^>]*id="${id}"[^>]*>([\\s\\S]*?)<\\/table>`));
  if (tableMatch) {
    tables.push(tableMatch[1]);
  }
});

const extractedData = [];

tables.forEach(tableContent => {
  // Split by <tr> but skip the header
  const trs = tableContent.split(/<tr[^>]*>/).slice(1);
  
  trs.forEach(tr => {
    // Address is inside an <a> tag
    const addrMatch = tr.match(/address\/([13][a-km-zA-HJ-NP-Z1-9]{25,34}|bc1[a-z0-9]{39,59})/);
    // Balance is in a <td> with data-val or just following the address
    // Looking for the first <td> after the address <td>
    // Actually, BitInfoCharts has data-val for the balance in the next <td>
    const balanceMatch = tr.match(/data-val="([\d,.]+)"[^>]*>([\d,.]+)\s*BTC/);
    
    if (addrMatch && balanceMatch) {
      extractedData.push({
        address: addrMatch[1],
        balance: balanceMatch[2] + ' BTC'
      });
    }
  });
});

if (extractedData.length > 0) {
  let md = "# Top Dormant Bitcoin Addresses (7+ Years)\n\n";
  md += "| Rank | Address | Balance |\n";
  md += "| :--- | :--- | :--- |\n";
  extractedData.slice(0, 100).forEach((item, index) => {
    md += `| ${index + 1} | ${item.address} | ${item.balance} |\n`;
  });
  fs.writeFileSync('dormant_addresses.md', md);
  console.log(`Successfully extracted ${Math.min(extractedData.length, 100)} addresses.`);
} else {
  console.error("No data extracted. The HTML structure might have changed.");
  // Debug: show a sample <tr>
  const firstTable = tables[0] || "";
  const firstTr = firstTable.match(/<tr[^>]*>[\s\S]*?<\/tr>/);
  if (firstTr) {
    console.log("Sample <tr> for debugging:", firstTr[0].slice(0, 500));
  }
}

File: extract_data.py
import os
import requests
from bs4 import BeautifulSoup
import pandas as pd
import time

def extract_dormant_btc():
    # Use the main list with offset to get more addresses
    # BitInfoCharts uses offset=100, 200, etc.
    base_url = "https://bitinfocharts.com/top-100-dormant_7y-bitcoin-addresses"
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
    }
    
    extracted_data = []
    
    for offset in range(0, 1000, 100):
        url = f"{base_url}-{offset // 100 + 1}.html" if offset > 0 else f"{base_url}.html"
        # Actually, BitInfoCharts top lists often use -2, -3 for pages, and each page has 100
        # BUT our previous attempt showed only 19? That's strange. 
        # Let's try to find ALL tables.
        
        print(f"Fetching: {url}...")
        
        try:
            response = requests.get(url, headers=headers, timeout=15)
            if response.status_code != 200:
                print(f"Failed: {response.status_code}")
                break
                
            soup = BeautifulSoup(response.text, 'html.parser')
            
            # They might have multiple tables or different IDs
            tables = soup.find_all('table')
            page_items = 0
            
            for table in tables:
                # Check if this looks like a data table
                rows = table.find_all('tr')
                for row in rows:
                    cols = row.find_all('td')
                    if len(cols) < 3: continue
                    
                    rank_text = cols[0].get_text(strip=True)
                    if not rank_text or not rank_text[0].isdigit():
                        continue
                    
                    rank = rank_text.rstrip('.')
                    address_cell = cols[1]
                    address_link = address_cell.find('a')
                    if not address_link: continue
                    address = address_link.get_text(strip=True)
                    
                    # Skip if not a valid BTC address (roughly)
                    if not (address.startswith('1') or address.startswith('3') or address.startswith('bc1')):
                        continue

                    label = ""
                    small_tag = address_cell.find('small')
                    if small_tag:
                        label = small_tag.get_text(strip=True)
                    
                    balance_text = cols[2].get_text(separator=" ", strip=True)
                    btc_balance = balance_text.split("BTC")[0].strip()
                    
                    extracted_data.append({
                        "Rank": rank,
                        "Address": address,
                        "Label": label,
                        "Balance": btc_balance + " BTC"
                    })
                    page_items += 1
            
            print(f"Found {page_items} items on this page.")
            if page_items == 0: break
            
            time.sleep(2)
            
        except Exception as e:
            print(f"Error: {e}")
            break

    if extracted_data:
        df = pd.DataFrame(extracted_data)
        df = df.drop_duplicates(subset=['Address'])
        df['Rank_Int'] = pd.to_numeric(df['Rank'], errors='coerce')
        df = df.sort_values('Rank_Int').drop(columns=['Rank_Int'])
        
        df.to_csv("dormant_addresses.csv", index=False)
        print(f"Total extracted: {len(df)}")
    else:
        print("Failed.")

if __name__ == "__main__":
    extract_dormant_btc()
